### ISSUE TITLE
An invalid response was received from the upstream server

### ISSUE THREAD

**Issue by Renz2018** [Kong/kubernetes-ingress-controller#358]
### Summary

**Kong Ingress controller version**

 Kong Ingress Controller 0.5.0
**Kong or Kong Enterprise version**
Kong: 1.2

### Kubernetes version

Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.6", GitCommit:"01a904e", GitTreeState:"", BuildDate:"2019-02-28T13:54:16Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}

### Environment

- **Cloud provider or hardware configuration**: aliyun
- **Install tools** : install kong using all-in-one YAMLs(pg db)
- **Others**: install apps(micro services) use helm

### What happened
when after helm upgrade some-apps or some pods killed and schedule other pod's ip, 
Can't access the app, kong return Kong Error: An invalid response was received from the upstream server, In konga's upstream page, I find the target is updated, But I look up kong pod's log I found 
```
31#0: *11140685 connect() failed (111: Connection refused) while connecting to upstream, client: 10.0.73.147, server: kong, request: "GET /v2/api-docs HTTP/1.1", upstream: "http://192.168.2.213:8082/v2/api-docs", host: "***-test.***.com"
```
Kong still use old ip(192.168.2.213:8082) to connect upstream
### Expected behvaior
kong should connect update pod ip

**Comment by hbagdi** [comment]
Hello @Renz2018,

Thanks for reporting the issue.
Could you share the timeline for these events relative to the upgrade?
How long after the upgrade do you see this issue?

There is a small delay between the time a Pod is terminated to the time it is actually removed from Kong's configuration. This exists because once the controller is notified of the. change, Kong's database is updated and then all other Kong nodes pick that change up.

How many pods of Kong are running?

**Comment by Renz2018** [comment]
1. I first found this problem(Kong Error: An invalid response was received from the upstream server) about 1-2 hours after the micro-service update(helm upgrade)
2. I check upstream in konga's upstream page,  target's ip is correct.
3. I check kong-ingress-controller(ingress-controller container) log, no error.
4.I can delete pod manully , and pod reschedule, then upstream can be correct.
Besause It's not 100% happening , so It's hard to test.
I use kubectl scale pods of kong to 2

**Comment by hbagdi** [comment]
This is another instance of https://github.com/Kong/kong/issues/4846.

Could you please try 1.4rc1 and see if it resolves your issue?

**Comment by adambom** [comment]
I'm having this problem too using version 1.4 of Kong. I'm getting logs in the proxy that look like this when attempting to hit the service:

```
2019/11/21 17:06:28 [error] 33#0: *15430227 connect() failed (111: Connection refused) while connecting to upstream, client: 18.208.181.247, server: kong, request: "GET / HTTP/1.1", upstream: "http://10.245.133.73:80/", host: "*.com"
```

It didn't happen at first, but it started happening hours after I created the ingress. It seems like the pods got rescheduled and Kong is using a stale IP address.

I'm also seeing logs in the ingress controller:

```
I1121 12:18:37.639654       1 status.go:201] new leader elected: kong-main-kong-6b57c9dc58-5jzgj
E1119 02:45:38.128522       1 reports.go:107] error sending report: write udp 10.249.132.220:45262->34.195.241.46:61829: write: connection refused
E1119 04:45:38.126994       1 reports.go:107] error sending report: write udp 10.249.132.220:45262->34.195.241.46:61829: write: connection refused
E1119 06:45:38.126979       1 reports.go:107] error sending report: write udp 10.249.132.220:45262->34.195.241.46:61829: write: connection refused
E1119 11:45:38.127018       1 reports.go:107] error sending report: write udp 10.249.132.220:45262->34.195.241.46:61829: write: connection refused
E1119 19:45:38.126992       1 reports.go:107] error sending report: write udp 10.249.132.220:45262->34.195.241.46:61829: write: connection refused
E1121 09:40:52.514862       1 leaderelection.go:324] error retrieving resource lock infrastructure/kong-ingress-controller-leader-kong-kong: Get https://10.249.0.1:443/api/v1/namespaces/infrastructure/configmaps/kong-ingress-controller-leader-kong-kong: dial tcp 10.249.0.1:443: connect: connection refused
E1121 09:56:09.173761       1 reflector.go:283] pkg/mod/k8s.io/client-go@v0.0.0-20190819141724-e14f31a72a77/tools/cache/reflector.go:98: Failed to watch *v1.KongIngress: Get https://10.249.0.1:443/apis/configuration.konghq.com/v1/kongingresses?resourceVersion=68133203&timeout=8m37s&timeoutSeconds=517&watch=true: dial tcp 10.249.0.1:443: connect: connection refused
E1121 09:56:09.173762       1 reflector.go:283] pkg/mod/k8s.io/client-go@v0.0.0-20190819141724-e14f31a72a77/tools/cache/reflector.go:98: Failed to watch *v1.KongPlugin: Get https://10.249.0.1:443/apis/configuration.konghq.com/v1/kongplugins?resourceVersion=68133203&timeout=5m33s&timeoutSeconds=333&watch=true: dial tcp 10.249.0.1:443: connect: connection refused
E1121 09:56:09.173761       1 reflector.go:283] pkg/mod/k8s.io/client-go@v0.0.0-20190819141724-e14f31a72a77/tools/cache/reflector.go:98: Failed to watch *v1.KongConsumer: Get https://10.249.0.1:443/apis/configuration.konghq.com/v1/kongconsumers?resourceVersion=68133203&timeout=5m59s&timeoutSeconds=359&watch=true: dial tcp 10.249.0.1:443: connect: connection refused
E1121 09:56:09.173769       1 reflector.go:283] pkg/mod/k8s.io/client-go@v0.0.0-20190819141724-e14f31a72a77/tools/cache/reflector.go:98: Failed to watch *v1beta1.Ingress: Get https://10.249.0.1:443/apis/extensions/v1beta1/ingresses?resourceVersion=67714204&timeout=8m16s&timeoutSeconds=496&watch=true: dial tcp 10.249.0.1:443: connect: connection refused
```

**Comment by hbagdi** [comment]
@adambom It seems like you are having a API-server outage or it is under heavy load and cannot respond (as seen from `connect: connection refused` error).

**Comment by adambom** [comment]
If that was the case, wouldn't the ingress controller recover when the outage was over? Other ingress rules continued to work even while this one was down.

**Comment by hbagdi** [comment]
You can ignore the errors from `reports.go safely.
Did the errors from `reflector.go` resolve itself after a while or is that persistent?

**Comment by adambom** [comment]
The errors were persistent and did not go away until I restarted the containers.

**Comment by hbagdi** [comment]
This seems like OP's error but the root cause is different. Can we not hijack this thread an open a new issue to continue the investigation there?
Please share the deployment spec of Kong in the new issue. Thanks @adambom!

**Comment by adambom** [comment]
Here is the issue: https://github.com/Kong/kubernetes-ingress-controller/issues/463

**Comment by hbagdi** [comment]
This should be fixed with Kong 1.4.2. Please re-open if the problem persists.
